Geometry/Fields/VectorFields/weightWindowsField_class.f90

```diff
@@ -0,0 +1,165 @@
+  !!   upperW ->  array with the weight windows upper weights;
+  !!   constSurvival ->  constant to be multiplied to lower weight for getting
+  !!                     roulette survival weight
+  !!
+  !! Interface:
+  !!   vectorField interface
+  !!
+  type, public, extends(vectorField) :: weightWindowsField
+    class(tallyMap), allocatable :: net
+    integer(shortInt)            :: N
+    real(defReal), dimension(:), allocatable :: lowerW
+    real(defReal), dimension(:), allocatable :: upperW
+    real(defReal) :: constSurvival = ZERO
+  contains
+    ! Superclass interface
+    procedure :: init
+    procedure :: kill
+    procedure :: at
+  end type weightWindowsField
+
+contains
+
+  !!
+  !! Initialise from dictionary
+  !!
+  !! See field_inter for details
+  !!
+  subroutine init(self, dict)
+    class(weightWindowsField), intent(inout) :: self
+    class(dictionary), intent(in)            :: dict
+    type(dictionary)                         :: dict2
+    character(pathLen)                       :: path
+    integer(shortInt), parameter  :: ALL = 0
+    character(100), parameter :: Here = 'init (weightWindowsField_class.f90)'
+
+    call dict % get(path,'file')
+
+    ! Load dictionary
+    call fileToDict(dict2, path)
+
+    ! Initialise overlay map
+    call new_tallyMap(self % net, dict2 % getDictPtr('map'))
+    self % N = self % net % bins(ALL)
+
+    ! Read weight values from file
+    call dict2 % get(self % constSurvival, 'constSurvival')
+    call dict2 % get(self % lowerW, 'wLower')
+    call dict2 % get(self % upperW, 'wUpper')
+
+    ! Check data consistency
+    if (size(self % lowerW) /= self % N) then
+      call fatalError(Here, 'Size of weight window net does not match the map provided')
+    end if
+
+    if (self % constSurvival <= ONE) call fatalError(Here, 'Survival constant must be bigger than one')
+    if (any(self % lowerW * self % constSurvival > self % upperW)) then
+      call fatalError(Here, 'Russian Roulette survival weight must be smaller &
+                           & than the weight window upper weight')
+    end if
+
+
+  end subroutine init
+
+  !!
+  !! Return to uninitialised state
+  !!
+  elemental subroutine kill(self)
+    class(weightWindowsField), intent(inout) :: self
+
+    call self % net % kill()
+    deallocate(self % net)
+
+  end subroutine kill
+
+  !!
+  !! Get value of the vector field given the phase-space location of a particle
+  !!
+  !! See vectorField_inter for details
+  !!
+  function at(self, p) result(val)
+    class(weightWindowsField), intent(in) :: self
+    class(particle), intent(inout)        :: p
+    real(defReal), dimension(3)           :: val
+    type(particleState)                   :: state
+    integer(shortInt)                     :: binIdx
+
+    ! Get current particle state
+    state = p
+
+    ! Read map bin index
+    binIdx = self % net % map(state)
+
+    ! Return if invalid bin index
+    if (binIdx == 0) then
```

Comment:
(Note to self) How happy are different parts of the code to receive vector of 0s as weight window parameters.

Reply:
It's ok, if the weight limits are zero simply nothing happens. Zero weights can happen if the map doesn't cover the whole geometry, or if some values in the weight windows are zero (it could happen if you produce your weight windows in an iterative way in MC, during the first iterations particles might not make it far away).

Reply:
Sorry! I made this comment as a note for myself when assembling a review, but I forgot to remove it!

Reply:
No problem, it is a fair point!